### DISCUSSION TITLE
How to define and use global exceptions？

### DISCUSSION THREAD

**Discussion by lnsane** [fastapi/fastapi#7373] [Questions] [answered]
```Python
import uvicorn
from fastapi import FastAPI
from flask import Request
from fastapi.responses import JSONResponse
from Api.api import api_router


class UnicornException(Exception):
    def __init__(self, name: str):
        self.name = name


app = FastAPI()


@app.exception_handler(UnicornException)
async def unicorn_exception_handler(request: Request, exc: UnicornException):
    return JSONResponse(
        status_code=418,
        content={"message": f"Oops! {exc.name} did something. There goes a rainbow..."},
    )


app.include_router(api_router, prefix="/api/v1")

if __name__ == "__main__":
    uvicorn.run("Start:app", host="0.0.0.0", port=8080, log_level="info")

```
```python3
from fastapi import APIRouter

from Api import webApi

api_router = APIRouter()
api_router.include_router(webApi.router, tags=["webApi"])

```
```python3
from fastapi import APIRouter

from Start import UnicornException
from WebInit import test
from allmain.StartInit import stop_all_containers, get_ssh_public_key, request_yml_data
from bean.YmalBean import Yml

router = APIRouter()


@router.get("/1")
def read_root():
    # StartInit.get_ssh_public_key()
    raise UnicornException(name="asdasd")
```
```bash
Traceback (most recent call last):
  File "/home/k/PycharmProjects/auto/Start.py", line 5, in <module>
    from Api.api import api_router
  File "/home/k/PycharmProjects/auto/Api/api.py", line 3, in <module>
    from Api import webApi
  File "/home/k/PycharmProjects/auto/Api/webApi.py", line 3, in <module>
    from Start import UnicornException
  File "/home/k/PycharmProjects/auto/Start.py", line 5, in <module>
    from Api.api import api_router
ImportError: cannot import name 'api_router'

```

**Comment by MacMacky** [comment, marked as answer]
You're having a __Circular Import__ problem. I suggest you do this.

__Start.py__ - under root folder
```py
import uvicorn
from fastapi import FastAPI
from flask import Request
from fastapi.responses import JSONResponse
from Api.api import api_router
from exceptions import main # or from exceptions.main import UnicornException

app = FastAPI()

@app.exception_handler(main.UnicornException)
async def unicorn_exception_handler(request: Request, exc: main.UnicornException):
    return JSONResponse(
        status_code=418,
        content={
            "message": f"Oops! {exc.name} did something. There goes a rainbow..."},
    )


app.include_router(api_router, prefix="/api/v1")

if __name__ == "__main__":
    uvicorn.run("Start:app", host="0.0.0.0", port=8080, log_level="info")
```

__Api/api.py__
```py
from fastapi import APIRouter

from Api import webApi

api_router = APIRouter()
api_router.include_router(webApi.router, tags=["webApi"])
```

__Api/webApi.py__
```py
from fastapi import APIRouter
from exceptions import main # or from exceptions.main import UnicornException

router = APIRouter()

@router.get("/1")
def read_root():
    raise main.UnicornException(name="Error")
```

__exceptions/main.py__
```py

class UnicornException(Exception):
    def __init__(self, name: str):
        self.name = name


class AnotherException(Exception):
    def __init__(self, message: str, status: int):
        self.message = message
        self.status = status
```

**Comment by lnsane** [comment]
> You're having a **Circular Import** problem. I suggest you do this.
> 
> **Start.py** - under root folder
> 
> ```python
> import uvicorn
> from fastapi import FastAPI
> from flask import Request
> from fastapi.responses import JSONResponse
> from Api.api import api_router
> from exceptions import main # or from exceptions.main import UnicornException
> 
> app = FastAPI()
> 
> @app.exception_handler(main.UnicornException)
> async def unicorn_exception_handler(request: Request, exc: main.UnicornException):
>     return JSONResponse(
>         status_code=418,
>         content={
>             "message": f"Oops! {exc.name} did something. There goes a rainbow..."},
>     )
> 
> 
> app.include_router(api_router, prefix="/api/v1")
> 
> if __name__ == "__main__":
>     uvicorn.run("Start:app", host="0.0.0.0", port=8080, log_level="info")
> ```
> 
> **Api/api.py**
> 
> ```python
> from fastapi import APIRouter
> 
> from Api import webApi
> 
> api_router = APIRouter()
> api_router.include_router(webApi.router, tags=["webApi"])
> ```
> 
> **Api/webApi.py**
> 
> ```python
> from fastapi import APIRouter
> from exceptions import main # or from exceptions.main import UnicornException
> 
> router = APIRouter()
> 
> @router.get("/1")
> def read_root():
>     raise main.UnicornException(name="Error")
> ```
> 
> **exceptions/main.py**
> 
> ```python
> class UnicornException(Exception):
>     def __init__(self, name: str):
>         self.name = name
> 
> 
> class AnotherException(Exception):
>     def __init__(self, message: str, status: int):
>         self.message = message
>         self.status = status
> ```

is so good! thanks

**Comment by tiangolo** [comment]
Thanks for the help here @MacMacky ! :clap: :bow: 

Thanks for reporting back and closing the issue @lnsane  :+1:
